tybalt/data_models.py

```diff
@@ -0,0 +1,245 @@
+                               index=self.df.index,
+                               columns=self.df.columns)
+
+    def pca(self, n_components, transform_df=False):
+        self.pca_fit = decomposition.PCA(n_components=n_components)
+        self.pca_fit = self.pca_fit.fit(self.df)
+        self.pca_df = self.pca_fit.transform(self.df)
+        colnames = ['pca_{}'.format(x) for x in range(0, n_components)]
+        self.pca_df = pd.DataFrame(self.pca_df, index=self.df.index,
+                                   columns=colnames)
+        if transform_df:
+            out_df = self.pca_fit.transform(transform_df)
+            return out_df
+
+    def ica(self, n_components, transform_df=False):
+        self.ica_fit = decomposition.FastICA(n_components=n_components)
+        self.ica_fit = self.ica_fit.fit(self.df)
+        self.ica_df = self.ica_fit.transform(self.df)
+        colnames = ['ica_{}'.format(x) for x in range(0, n_components)]
+        self.ica_df = pd.DataFrame(self.ica_df, index=self.df.index,
+                                   columns=colnames)
+        if transform_df:
+            out_df = self.ica_fit.transform(transform_df)
+            return out_df
+
+    def nmf(self, n_components, transform_df=False, init='nndsvdar', tol=5e-3):
+        self.nmf_fit = decomposition.NMF(n_components=n_components, init=init,
+                                         tol=tol)
+        self.nmf_fit = self.nmf_fit.fit(self.df)
+        self.nmf_df = self.nmf_fit.transform(self.df)
+        colnames = ['nmf_{}'.format(x) for x in range(0, n_components)]
+        self.nmf_df = pd.DataFrame(self.nmf_df, index=self.df.index,
+                                   columns=colnames)
+        if transform_df:
+            out_df = self.nmf_fit.transform(transform_df)
+            return out_df
+
+    def nn(self, n_components, model='tybalt', transform_df=False, **kwargs):
+        # unpack kwargs
+        original_dim = kwargs.pop('original_dim', self.df.shape[1])
+        latent_dim = kwargs.pop('latent_dim', n_components)
+        batch_size = kwargs.pop('batch_size', 50)
+        epochs = kwargs.pop('epochs', 50)
+        learning_rate = kwargs.pop('learning_rate', 0.0005)
+        noise = kwargs.pop('noise', 0)
+        sparsity = kwargs.pop('sparsity', 0)
+        kappa = kwargs.pop('kappa', 1)
+        epsilon_std = kwargs.pop('epsilon_std', 1.0)
+        beta = kwargs.pop('beta', 0)
+        beta = K.variable(beta)
+        loss = kwargs.pop('loss', 'binary_crossentropy')
+        validation_ratio = kwargs.pop('validation_ratio', 0.1)
+        if model == 'ctybalt':
+            y_df = kwargs.pop('y_df', self.other_df)
+            y_var = kwargs.pop('y_var', 'groups')
+            label_dim = kwargs.pop('label_dim', len(set(y_df[y_var])))
+
+        self.nn_test_df = self.df.sample(frac=validation_ratio)
+        self.nn_train_df = self.df.drop(self.nn_test_df.index)
+
+        if hasattr(self, 'other_df'):
```

Comment:
What is going on here? I'm a little confused why you have condition here. Does this align with the ctybalt condition above?

Reply:
yeah, nice catch. This is an error. I will update in the next commit!